### PR TITLE
Named vars

### PR DESCRIPTION
**Pull request by lucianojoublanc-da**
TL; DR
---------

We've changed the encoding of `Observation` so that it admits a `Functor` instance now.

Requirements
-------------------

Let's take a concrete example to explain what this is for.

Currently, if we want to model an options chain, (i.e. options with the same maturity but different strike prices, from low to high), we are required to create n trees of `Claim`s, with _n_ being the number of strikes we want to model.
Ideally, what we'd want to support, is the ability to create claims that have parameters. i.e. a single claim representing an 'option', and then we can use that to either create multiple instances of that option, but with a different strike price.

Solution
-----------

The approach I've taken is to re-model the `Observation` type to make it a proper `Functor`. What this means is that we can now create `Claim`s (by extension of `Observable`s) that have an arbitrary type in the leaves, not just a `Decimal` value.

How does this help us? It means that we can model a 'template', by letting the value type be a `Text`, an identifier name, instead of it's actual decimal value. So instead of creating a `Claim` with it's strike price `99.0`, we have a `Claim` with it's strike price `"strike"`, i.e. a string. This is our template.

Once we want to 'instantiate our template', we simply `fmap` those `Observation`s from their placeholder parameter names (`Text`), to their actual values (`Decimal`). This is trivial to do by passing a `Map` holding those instantiation parameters.
This so-called 'instantiation' can be either done at instrument issue time, or it can be deferred until evaluation; in other words, you have the option of storing your claim instances as `Claim f t Decimal a`, or as a pair of `(Map Text Decimal, Claim f t Text a)`.

Another subtlety which I want to elaborate on here is that of how time is represented in a template vs an instance of it. This is not an enhancement for this PR - it's already there - but I want to bring this to your attention.
When creating a template, I suggest it is stored as `Claim f Int Text a` . i.e. we store dates/times as an ordinal number, with zero being issue date. In our running example, this saves us from having to create multiple templates for the same option but different maturities. We can have a single option, where the maturity is in 30 days time.
Once the option is instantiated, we (contra)map the ordinal value into a `Date` or timestamp.

Implementation
--------------------

A summary of the changes:
- `Observation` signature remains unchanged
- However, inequality operations have been moved into a separate type `Observable.Inequality`
- In fact there is now a single inequality operator `>=`. Equality has been removed. This isn't as restrictive as it sounds - most `Observable`s will represent stochastic processes i.e. probabilities, so the main operation is `<=` rather than `==`. 
- Claim has an extra type parameter, which is passed-through to the `Observable` output parameter.

**Changes to `Observation`**
I've managed to replace the unholy mess on the left-hand side with a much simpler and cleaner implementation.
In the old version, `a` was a phantom type. Now it's actually a concrete value. Although we always envisaged the use of this DSL inside DAML, The advantage of this is that users can now write much safer `Observation`s off-ledger. Degenerate cases are caught at compile time (if the off-ledger language is strongly typed), or at worst at deserialization time - it never makes it onto the ledger.
```diff
 
 -- | Concrete implementation of `Observable`, which can be serialized.
-data Observation t a
-  = DecimalConst Decimal
-  | DecimalObs Key
-  | DecimalLte (Observation t Decimal, Observation t Decimal)
-  | DecimalEqu (Observation t Decimal, Observation t Decimal)
-  | DecimalAdd (Observation t Decimal, Observation t Decimal)
-  | DecimalNeg (Observation t Decimal)
-  | DecimalMul (Observation t Decimal, Observation t Decimal)
-  | DecimalDiv (Observation t Decimal, Observation t Decimal)
-  | BoolConst Bool
-  | DateIdentity
-  | DateConst t
-  | DateLte (Observation t t, Observation t t)
-  | DateEqu (Observation t t, Observation t t)
+-- Conceptually it's helpfult to think of this as the type `t -> x`, or `t -> Update x`.
+data Observation t x
+  = Const with value: x 
+    -- ^ A named constant
+  | Observe with key: O.Key 
+    -- ^ A named parameter
+  | Identity 
+    -- ^ The identity function, `t -> t`; useful for time comparisons.
+    -- e.g. `Inequality(Identity, Const "tomorrow" (date _))` equates to `time <= tomorrow`
+  | Add (Observation t x, Observation t x)
+  | Neg (Observation t x)
+  | Mul (Observation t x, Observation t x)
+  | Div (Observation t x, Observation t x)
   deriving (Eq, Show)
 ```
Below is an example of how this simplifies the algebra, having a `Functor` instance. Previously we relied on typeclass tricks in order to prevent users from creating and evaluating invalid `Observation`s. You can see that 'degenerate' cases trigger failure in `eval _ _ = error` in the old approach. Now those valid cases can't even be created anymore (ok, there are still some exceptions, but it's _much_ safe now!).

```diff
-class Interpret t a where
-  eval : Monad m => (Key -> t -> m Decimal) -> Observation t a -> Kleisli m t a
-
-instance Ord t => Interpret t Bool where
-  eval _ (BoolConst b) = pure b
-  eval spot (DecimalLte (d, d')) = liftA2 (<=) (eval spot d) (eval spot d')
-  eval spot (DecimalEqu (d, d')) = liftA2 (==) (eval spot d) (eval spot d')
-  eval spot (DateLte (d, d')) = liftA2 (<=) (eval spot d) (eval spot d')
-  eval spot (DateEqu (d, d')) = liftA2 (==) (eval spot d) (eval spot d')
-  eval _ _ = error "eval: Interpret Bool"
-
-instance Interpret t t where
-  eval _ (DateConst t) = pure t
-  eval _ DateIdentity = id
-  eval _ _ = error "eval: Interpret Date"
-
-instance Interpret t Decimal where
-  eval _ (DecimalConst d) = pure d
-  eval spot (DecimalObs key) = Kleisli $ spot key
-  eval spot (DecimalAdd (x, x')) = liftA2 (+) (eval spot x) (eval spot x')
-  eval spot (DecimalNeg x) = fmap (\x -> (- x)) (eval spot x)
-  eval spot (DecimalMul (x, x')) = liftA2 (*) (eval spot x) (eval spot x')
-  eval spot (DecimalDiv (x, x')) = liftA2 (/) (eval spot x) (eval spot x')
-  eval _ _ = error "eval: Interpret Decimal"
+eval : (Divisible x, Number x, Monad m) 
+     => (t -> x)
+     -> (O.Key -> t -> m x) 
+     -> Observation t x 
+     -> Kleisli m t x
+eval yoneda doObserve = cataM \case
+  ConstF x -> pure x
+  ObserveF key -> Kleisli (doObserve key)
+  AddF (x, x') -> pure $ x Prelude.+ x'
+  NegF x -> pure $ negate x
+  MulF (x, x') -> pure $ x * x'
+  DivF (x, x') -> pure $ x / x'
+  IdentityF -> arr yoneda
```

**Changes to `Claim`**

We now pass-thru the output type of `Observable`, so it becomes visible in `Claim`. Having one more parameter increases complexity, but this is the trade-off.

```diff
-data Claim f t a
+data Claim f t x a
```

There is also a new type `Inequality`. This is require to make the new `Observation` algebra sound, because the signature of inequalities (`a -> a -> Bool`) is different from that of ring operations (`a -> a -> a`).

```haskell
-- | Think of this as `f t t ≤ f t t | f t x ≤ f t x`
data Inequality f t x = LteT (f t t, f t t) | LteX (f t x, f t x)
```
And it's used here:
```diff
-  | When with predicate: f t Bool, claim: Claim f t a
+  | When with predicate: Inequality f t x, claim: Claim f t x a
```

### CODE REVIEW

File: daml/ContingentClaims/Observation.daml [hunk 1]
```diff
@@ -7,140 +7,103 @@
...
+import Daml.Control.Recursion (Recursive(..), Corecursive(..))
+import DA.Traversable (Traversable(..))
+import DA.Foldable (Foldable(..))
+import Prelude hiding (key, (.))
 
-type Key = Text
 type T = Observation
 
 -- | Concrete implementation of `Observable`, which can be serialized.
-data Observation t a
```
Comment (lucianojoublanc-da) [inline daml/ContingentClaims/Observation.daml:25]: I've managed to replace the unholy mess on the left-hand side with a much simpler and cleaner implementation.
In the old version, `a` was a phantom type. Now it's actually a concrete value. Although we always envisaged the use of this DSL inside DAML, The advantage of this is that users can now write much safer `Observation`s off-ledger. Degenerate cases are caught at compile time (if the off-ledger language is strongly typed), or at worst at deserialization time - it never makes it onto the ledger.

**Comment by lucianojoublanc-da** [comment]
Worth mentioning that one other approach I considered was to have everything stored as a `Decimal`. This was a suggestion we had from a conversation with Sofus. Here's an example of what he did:
https://github.com/sofusmortensen/BP-DAML/blob/5877dfc9caca4c35359dee9925627e9106e85ea6/daml/Kontract.daml#L44

I went against this eventually as we can't serialize types, so inspecting the ledger would have been much more difficult.

**Comment by lucianojoublanc-da** [comment]
> Ideally, what we'd want to support, is the ability to create claims that have parameters
And to be clear - we _can_ do this currently. Simply write a function.

To be more precise, we want a way of being able to reify this, e.g. serialize it on ledger so someone can inspect this function by looking at the data.

File: daml/ContingentClaims/Observation.daml [hunk 2]
```diff
@@ -7,140 +7,103 @@
...
-  | DecimalAdd (Observation t Decimal, Observation t Decimal)
-  | DecimalNeg (Observation t Decimal)
-  | DecimalMul (Observation t Decimal, Observation t Decimal)
-  | DecimalDiv (Observation t Decimal, Observation t Decimal)
-  | BoolConst Bool
-  | DateIdentity
-  | DateConst t
-  | DateLte (Observation t t, Observation t t)
-  | DateEqu (Observation t t, Observation t t)
+-- Conceptually it's helpfult to think of this as the type `t -> x`, or `t -> Update x`.
```
Comment (richardkapolnai-da) [inline daml/ContingentClaims/Observation.daml:24]:
```suggestion
-- Conceptually it's helpful to think of this as the type `t -> x`, or `t -> Update x`.
```

**Comment by leonidr-da** [comment]
I think that we need to be able to name the calculation within an observation not just the a constant or key. In FXO traders talk about strikes in terms of delta's (30 delta string on EuroDollar). So the strike would be 30 delta.

Furthermore, we would want to have names for assets as well. A Call can be on lots of different stocks, bonds, ...etc.

**Comment by lucianojoublanc-da** [comment]
> Furthermore, we would want to have names for assets as well.

This we can already do. It's captured in the `a` parameter in a `Claim`. For example, create

```haskell
data Asset = Asset with
  isin : ISIN
  description : Text
  claims : Claim Observation Date Asset
```

You can also use this approach to lookup assets based on structure, by keying your templates on `Claim`. That's what we've essentially done in https://github.com/digital-asset/da-marketplace/blob/a0879bf9160239d9767cf2ff0f8fac774d7280b7/daml/Marketplace/Issuance/AssetDescription.daml .

File: daml/ContingentClaims/Observable.daml [hunk 3]
```diff
@@ -3,45 +3,27 @@
...
+
+deriving instance (Eq t, Eq (f t x)) => Eq (Inequality f t x)
+
+deriving instance (Show t, Show (f t x)) => Show (Inequality f t x)
+
+class Point f t x where
+  pure : x -> f t x
+  observe : Key -> f t x
+
+class Interpret f where
```
Comment (lucianojoublanc-da) [inline daml/ContingentClaims/Observable.daml:27]: This needs a better name. When we add other interpreters, what will we call them?

`eval` is unsafe - it will fail when we try to lookup an Observable that is not yet known.
Compare this with valuation of the claim, which could also be considered `eval`uation, but which doesn't fail.

File: daml/ContingentClaims/Observation.daml [hunk 4]
```diff
@@ -7,140 +7,111 @@
...
-  | DateIdentity
-  | DateConst t
-  | DateLte (Observation t t, Observation t t)
-  | DateEqu (Observation t t, Observation t t)
+-- Conceptually it's helpful to think of this as the type `t -> x`, or `t -> Update x`.
+data Observation t x
+  = Const with value: x 
+    -- ^ A named constant
+  | Observe with key: O.Key 
+    -- ^ A named parameter
```
Comment (lucianojoublanc-da) [inline daml/ContingentClaims/Observation.daml:28]:
```suggestion
data Observation t a x
  = Param (Either a x )
    -- ^ A named parameter
```

This would have two advantages: 
1. We'd be able to write proper bi-functor instances for Observable.
2. We could tie the key to the asset at the `Claim` level. The reason I've avoided that and used a `type Key = Text` so far, is that as @georg-da pointed out, we might want to distinguish between 'close', 'high' 'low', etc. (but this could be worked around by wrapping the asset in an effect, that specifies what type of data to fetch). Also, not all parameters are necessarily 'assets' e.g. temperature etc.

File: daml/ContingentClaims/Claim.daml [hunk 5]
```diff
@@ -19,34 +20,52 @@ type F = ClaimF
...
+-- | Used to model cashflows of instruments.
 -- See [quickstart](../QUICKSTART.md) for detailed explanation.
 -- In Petyon-Jones' paper, this is called 'Contract'.
--- We renamed it to avoid ambiguity
--- * `f` is the effect, e.g. `(->)`
--- * `t` is time e.g. `Date`
--- * `a` is the asset
-data Claim f t a
+-- We renamed it to avoid ambiguity.
+-- * `f`, `t` and `x` correspond to the `Observable`, it's input and output, respectively.
```
Comment (brian-weir) [inline daml/ContingentClaims/Claim.daml:27]:
```suggestion
-- * `f`, `t` and `x` respectively correspond to the `Observable`, it's input type and the resulting output type.
```

Comment (brian-weir) [reply daml/ContingentClaims/Claim.daml:27]: Trying to make this statement a little clearer.

File: daml/ContingentClaims/Claim.daml [hunk 6]
```diff
@@ -19,34 +20,52 @@ type F = ClaimF
...
       -- ^ Used to combine multiple rights together.
-  | Or with lhs: Claim f t a, rhs: Claim f t a
+  | Or with lhs: Claim f t x a, rhs: Claim f t x a
       -- ^ Gives the bearer the right to choose between two claims.
-  | Cond with predicate: f t Bool, success: Claim f t a, failure: Claim f t a
+  | Cond with predicate: Inequality f t x, success: Claim f t x a, failure: Claim f t x a --TODO rename predicate -> `lte` or `inequality`
       -- ^ Gives the bearer the right to the first claim if `predicate` is true, else the second claim.
-  | Scale with k: f t Decimal, claim: Claim f t a
+  | Scale with k: f t x, claim: Claim f t x a
       -- ^ Multiplies the `claim` by `k` (which can be non-detrerministic).
```
Comment (brian-weir) [inline daml/ContingentClaims/Claim.daml:43]:
```suggestion
      -- ^ Multiplies the `claim` by `k` (which can be non-deterministic).
```

**Comment by chrisnorris-da** [comment]
Separately I would urge building out test coverage. Currently the metrics in a daml test run are just templates:33%, choices 0%.
just wrapping the mkContract in something to Clear them and check the results would increase the choices coverage to 20%.
Concretely:

File: daml/ContingentClaims/Observable.daml [hunk 7]
```diff
@@ -3,45 +3,41 @@
 -- SPDX-License-Identifier: Apache-2.0
 --
 
-{-# LANGUAGE NoImplicitPrelude #-}
+{-# LANGUAGE UndecidableInstances #-}
```
Comment (GeorgSchneider) [inline daml/ContingentClaims/Observable.daml:7]: Do we know if this pragma causes and issues, or is discouraged? In terms of (data-)dependencies or otherwise

File: daml/ContingentClaims/ObservableFunctions.daml [hunk 8]
```diff
@@ -9,32 +9,33 @@ module ContingentClaims.ObservableFunctions where
...
+  aunit = Prelude.pure Prelude.aunit
   (+) = liftA2 (Prelude.+)
-  negate = fmap (Prelude.negate)
+  negate = fmap Prelude.negate
+
+instance (Multiplicative a, Applicative (f t)) => Multiplicative (f t a) where
+  munit = Prelude.pure Prelude.munit
   (*) = liftA2 (Prelude.*)
-  (/) = liftA2 (Prelude./)
+  (^) = undefined
```
Comment (GeorgSchneider) [inline daml/ContingentClaims/ObservableFunctions.daml:33]: Why is this undefined, is there no instance to lift?

File: daml/ContingentClaims/Observable.daml (hunk 7 above)
Comment (lucianojoublanc-da) [reply daml/ContingentClaims/Observable.daml:7]: I'm not sure. But I've removed and everything seems to work :laughing: .

File: daml/ContingentClaims/Recursion.daml [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+
+apoM : (Monad m, Traversable f, Corecursive b f) => (a -> m (f (Either b a))) -> a -> m b
+apoM f a = (fmap embed <<< (>>= (mapA (pure ||| apoM f))) <<< f) a
+
+-- | Specialised lazy re-fold, used by `lifecycle`.
+apoCataM : (Monad m, Traversable f, Corecursive b f) => (f b -> b) -> (a -> m (f (Either b a))) -> a -> m b
+apoCataM g f a = _apoCataM g f a 
+
+-- | HIDE
+_apoCataM g f a = (fmap g <<< (>>= (mapA (pure ||| apoCataM g f))) <<< f) a
```
Comment (GeorgSchneider) [inline daml/ContingentClaims/Recursion.daml:48]: why is this helper function needed?

File: daml/ContingentClaims/ObservableFunctions.daml (hunk 8 above)
Comment (lucianojoublanc-da) [reply daml/ContingentClaims/ObservableFunctions.daml:33]: Good catch; I was just being lazy.

I've also fixed this for `Observation`, but only for powers 0,1, 2. Otherwise, we run the risk of creating super nested trees (as I've just implemented as unfolding `Mul` _n_ times). Not sure if that will be an issue.

File: daml/ContingentClaims/Recursion.daml (hunk 9 above)
Comment (lucianojoublanc-da) [reply daml/ContingentClaims/Recursion.daml:48]: It's not. I think it was left-over from something different I was trying to implement. :+1:

**Review by brian-weir — APPROVED** [review]
Looks good @lucianojoublanc-da !

I'm a bit confused as to why we're limiting `predicates` in the way we do in this PR despite your comment about it but I don't want to hold up the merge.

**Comment by lucianojoublanc-da** [comment]
> I'm a bit confused as to why we're limiting `predicates` in the way we do in this PR

As per our discussion, now the only available predicates are:

- comparing current time to a constant
- comparing one observable to another

So you're right, this is more constrained than before. It's going in the opposite direction of what e.g. #13 . The argument here is two fold. First, some of constraints can be expressed at the `Claim` level with the new operators we're introducing in #15 ; Second, from a design point of view, we want to keep the encoding of the `Claim/Inequality/Observable` managable. We need to remember that this is passed through the API, so the smaller the ADT is, the easier it is to manage. In other words, it's easier to 'add stuff later', rather than 'remove it'.
